### PR TITLE
Refactor core/Controller

### PR DESCRIPTION
**Pull request by shdehnavi**
## Summary
The required adjustments have been made to the classes in `/core/Controller ` namespace.

The improvements:

- Adding return types
- Adding throws types
- Adding types to properties
- Replacing qualifiers with imports
- Updating doc blocks
- Using PHP8's `str_starts_with` and `str_ends_with` instead of `substr`

## Checklist

- Code is [properly formatted](https://docs.nextcloud.com/server/latest/developer_manual/digging_deeper/continuous_integration.html#linting)
- [Sign-off message](https://github.com/src-d/guide/blob/master/developer-community/fix-DCO.md) is added to all commits

### CODE REVIEW

File: core/Controller/LoginController.php [hunk 1]
```diff
@@ -113,10 +114,10 @@ public function logout() {
 	 * @PublicPage
 	 * @NoCSRFRequired
 	 *
-	 * @param string $user
-	 * @param string $redirect_url
+	 * @param string|null $user
+	 * @param string|null $redirect_url
 	 *
-	 * @return TemplateResponse|RedirectResponse
+	 * @return Response
```
[the remark this answers is not in the document]
Comment (shdehnavi) [reply core/Controller/LoginController.php]: You are right, It's addressed in the next commit

File: core/Controller/PreviewController.php [hunk 2]
```diff
@@ -54,7 +58,15 @@ public function __construct(
 	 * @NoAdminRequired
 	 * @NoCSRFRequired
 	 *
-	 * @return DataResponse|FileDisplayResponse
+	 * @param string $file
+	 * @param int $x
+	 * @param int $y
+	 * @param bool $a
+	 * @param bool $forceIcon
+	 * @param string $mode
```
[the remark this answers is not in the document]
Comment (shdehnavi) [reply core/Controller/PreviewController.php]: It's addressed in the next commit

File: core/Controller/PreviewController.php [hunk 3]
```diff
@@ -81,15 +93,23 @@ public function getPreview(
...
+	 * @throws NotPermittedException
 	 */
 	public function getPreviewByFileId(
 		int $fileId = -1,
 		int $x = 32,
 		int $y = 32,
 		bool $a = false,
 		bool $forceIcon = true,
-		string $mode = 'fill') {
+		string $mode = 'fill'): DataResponse|Response|FileDisplayResponse {
```
[the remark this answers is not in the document]
Comment (shdehnavi) [reply core/Controller/PreviewController.php]: it's addressed in the next commit

File: core/Controller/ClientFlowLoginController.php [hunk 4]
```diff
@@ -224,11 +227,15 @@ public function grantPage(string $stateToken = '',
...
+	 * @param string $stateToken
+	 * @param string $clientIdentifier
+	 * @return Response|StandaloneTemplateResponse|RedirectResponse
+	 * @throws ClientNotFoundException
+	 * @throws Exception
 	 */
 	#[UseSession]
 	public function generateAppPassword(string $stateToken,
-										string $clientIdentifier = '') {
+										string $clientIdentifier = ''): Response|StandaloneTemplateResponse|Http\RedirectResponse {
```
Comment (nickvergessen) [inline core/Controller/ClientFlowLoginController.php:238]:
```suggestion
										string $clientIdentifier = ''): Response|StandaloneTemplateResponse|RedirectResponse {
```

File: core/Controller/ContactsMenuController.php [hunk 5]
```diff
@@ -54,10 +54,12 @@ public function index(?string $filter = null): array {
 	/**
 	 * @NoAdminRequired
 	 *
+	 * @param int $shareType
+	 * @param string $shareWith
 	 * @return JSONResponse|\JsonSerializable
```
Comment (nickvergessen) [inline core/Controller/ContactsMenuController.php:59]:
```suggestion
```

File: core/Controller/CssController.php [hunk 6]
```diff
@@ -64,7 +64,8 @@ public function __construct(
 	 *
 	 * @param string $fileName css filename with extension
 	 * @param string $appName css folder name
-	 * @return FileDisplayResponse|NotFoundResponse
+	 * @return Response
+	 * @throws \Exception
 	 */
 	public function getCss(string $fileName, string $appName): Response {
```
Comment (nickvergessen) [inline core/Controller/CssController.php:70]:
```suggestion
	 * @return FileDisplayResponse|NotFoundResponse
	 */
	public function getCss(string $fileName, string $appName): FileDisplayResponse|NotFoundResponse {
```

File: core/Controller/JsController.php [hunk 7]
```diff
@@ -64,7 +64,8 @@ public function __construct(
 	 *
 	 * @param string $fileName js filename with extension
 	 * @param string $appName js folder name
-	 * @return FileDisplayResponse|NotFoundResponse
+	 * @return Response
+	 * @throws \Exception
 	 */
 	public function getJs(string $fileName, string $appName): Response {
```
Comment (nickvergessen) [inline core/Controller/JsController.php:70]:
```suggestion
	 * @return FileDisplayResponse|NotFoundResponse
	 */
	public function getJs(string $fileName, string $appName): FileDisplayResponse|NotFoundResponse {
```

File: core/Controller/LoginController.php [hunk 8]
```diff
@@ -264,6 +265,12 @@ private function generateRedirect(?string $redirectUrl): RedirectResponse {
 	 * @NoCSRFRequired
 	 * @BruteForceProtection(action=login)
 	 *
+	 * @param Chain $loginChain
+	 * @param string $user
+	 * @param string $password
+	 * @param string|null $redirect_url
+	 * @param string $timezone
+	 * @param string $timezone_offset
 	 * @return RedirectResponse
```
Comment (nickvergessen) [inline core/Controller/LoginController.php:274]:
```suggestion
```

File: core/Controller/PreviewController.php [hunk 9]
```diff
@@ -81,15 +87,23 @@ public function getPreview(
...
 	 *
+	 * @param int $fileId
+	 * @param int $x
+	 * @param int $y
+	 * @param bool $a
+	 * @param bool $forceIcon
+	 * @param string $mode
 	 * @return DataResponse|FileDisplayResponse
+	 * @throws NoUserException
+	 * @throws NotPermittedException
```
Comment (nickvergessen) [inline core/Controller/PreviewController.php:98]:
```suggestion
```

File: core/Controller/PreviewController.php [hunk 10]
```diff
@@ -107,7 +121,15 @@ public function getPreviewByFileId(
...
-	 * @return DataResponse|FileDisplayResponse
+	 * @param Node $node
+	 * @param int $x
+	 * @param int $y
+	 * @param bool $a
+	 * @param bool $forceIcon
+	 * @param string $mode
+	 * @return Response
+	 * @throws NotFoundException
+	 * @throws InvalidPathException
```
Comment (nickvergessen) [inline core/Controller/PreviewController.php:132]:
```suggestion
	 * @return DataResponse|FileDisplayResponse
```

or you remove the whole doc block and change the end of the signature to:
```
string $mode) : DataResponse|FileDisplayResponse {
```

File: core/Controller/ClientFlowLoginController.php [hunk 11]
```diff
@@ -224,11 +227,15 @@ public function grantPage(string $stateToken = '',
 	/**
 	 * @NoAdminRequired
 	 *
-	 * @return Http\RedirectResponse|Response
+	 * @param string $stateToken
+	 * @param string $clientIdentifier
+	 * @return Response|StandaloneTemplateResponse|RedirectResponse
```
Comment (artonge) [inline core/Controller/ClientFlowLoginController.php:232]: As nickvergessen said, 

> usually we remove types from the doc block when they are reflected in the methods signature as they are just error prone to be missed on updates.

So you can remove the doc block comments for all methods that now have their type in their signature :)

File: core/Controller/ProfileApiController.php [hunk 12]
```diff
@@ -53,6 +54,14 @@ public function __construct(
...
 	 * @PasswordConfirmationRequired
 	 * @UserRateThrottle(limit=40, period=600)
+	 * @param string $targetUserId
+	 * @param string $paramId
+	 * @param string $visibility
+	 * @return DataResponse
+	 * @throws OCSBadRequestException
+	 * @throws OCSForbiddenException
+	 * @throws OCSNotFoundException
+	 * @throws Exception
```
Comment (nickvergessen) [inline core/Controller/ProfileApiController.php:64]:
```suggestion
	 * @throws OCSBadRequestException
	 * @throws OCSForbiddenException
	 * @throws OCSNotFoundException
```

The other things should not be needed as they are indicated by the method signature or not planned to happen/be catched (general exception)

File: core/Controller/TwoFactorChallengeController.php [hunk 13]
```diff
@@ -85,8 +85,9 @@ private function splitProvidersAndBackupCodes(array $providers): array {
 	 *
 	 * @param string $redirect_url
 	 * @return StandaloneTemplateResponse
+	 * @throws \Exception
```
Comment (nickvergessen) [inline core/Controller/TwoFactorChallengeController.php:88]:
```suggestion
```
Not part of the API.

File: core/Controller/TwoFactorChallengeController.php [hunk 14]
```diff
@@ -112,9 +113,10 @@ public function selectChallenge($redirect_url) {
 	 * @param string $challengeProviderId
 	 * @param string $redirect_url
 	 * @return StandaloneTemplateResponse|RedirectResponse
+	 * @throws \Exception
```
Comment (nickvergessen) [inline core/Controller/TwoFactorChallengeController.php:116]:
```suggestion
```
Not part of the API.

File: core/Controller/TwoFactorChallengeController.php [hunk 15]
```diff
@@ -204,6 +206,7 @@ public function solveChallenge($challengeProviderId, $challenge, $redirect_url =
 	/**
 	 * @NoAdminRequired
 	 * @NoCSRFRequired
+	 * @throws \Exception
```
Comment (nickvergessen) [inline core/Controller/TwoFactorChallengeController.php:209]:
```suggestion
```
Not part of the API.

File: core/Controller/TwoFactorChallengeController.php [hunk 16]
```diff
@@ -220,8 +223,9 @@ public function setupProviders(): StandaloneTemplateResponse {
...
```
Comment (nickvergessen) [inline core/Controller/TwoFactorChallengeController.php:226]:
```suggestion
```
Not part of the API.

File: core/Controller/WellKnownController.php [hunk 17]
```diff
@@ -44,6 +44,7 @@ public function __construct(
 	 * @PublicPage
 	 * @NoCSRFRequired
 	 *
+	 * @param string $service
 	 * @return Response
```
Comment (nickvergessen) [inline core/Controller/WellKnownController.php:48]:
```suggestion
```

**Review by nickvergessen — CHANGES_REQUESTED** [review]

File: core/Controller/ClientFlowLoginController.php (hunk 11 above)
Comment (shdehnavi) [reply core/Controller/ClientFlowLoginController.php:232]: It's addressed in the commit

File: core/Controller/LoginController.php [hunk 18]
```diff
@@ -45,6 +45,7 @@
 use OCP\AppFramework\Http\Attribute\UseSession;
 use OCP\AppFramework\Http\DataResponse;
 use OCP\AppFramework\Http\RedirectResponse;
+use OCP\AppFramework\Http\Response;
```
Comment (nickvergessen) [inline core/Controller/LoginController.php:48]: I don't see this being used in the diff. Is it an unused import?

File: core/Controller/PreviewController.php (hunk 10 above)
Comment (nickvergessen) [reply core/Controller/PreviewController.php:132]: See this

File: core/Controller/PreviewController.php (hunk 9 above)
Comment (nickvergessen) [reply core/Controller/PreviewController.php:98]: See this

File: core/Controller/ProfileApiController.php (hunk 12 above)
Comment (nickvergessen) [reply core/Controller/ProfileApiController.php:64]: another open comment here

File: core/Controller/TwoFactorChallengeController.php [hunk 19]
```diff
@@ -164,11 +164,11 @@ public function showChallenge($challengeProviderId, $redirect_url) {
 	 *
 	 * @param string $challengeProviderId
 	 * @param string $challenge
-	 * @param string $redirect_url
+	 * @param string|null $redirect_url
 	 * @return RedirectResponse
 	 */
 	#[UseSession]
-	public function solveChallenge($challengeProviderId, $challenge, $redirect_url = null) {
+	public function solveChallenge(string $challengeProviderId, string $challenge, string $redirect_url = null): RedirectResponse {
```
Comment (nickvergessen) [inline core/Controller/TwoFactorChallengeController.php:171]:
```suggestion
	public function solveChallenge(string $challengeProviderId, string $challenge, ?string $redirect_url = null): RedirectResponse {
```

**Review by nickvergessen — CHANGES_REQUESTED** [review]

**Pull request closed without merging**